### DISCUSSION TITLE
Object Lifecycle Events with Async Sessions/Sessionmakers

### DISCUSSION THREAD

**Discussion by stanleyfang** [sqlalchemy/sqlalchemy#12079] [Usage Questions] [answered]
Hello, I'm currently using a 'transient_to_pending' object lifecycle event to handle updates to a handful of many-to-many relations I have in the database (pretty much exactly how it's been done in the 'UniqueObjectValidatedOnPending' recipe here: https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObjectValidatedOnPending).

I'm in the middle of a switch from sync to async, but I've run into an issue: this object lifecycle event doesn't seem to play well with the documented usage of events with async sessions. The documented usage of a 'before_commit' event works fine however.

e.g.
```
def my_before_commit(session: Session) -> None:
    session.execute(text("SELECT 1"))
    print("Before commit")


def my_transient_to_pending(session: Session, object_: Base) -> None:
    session.execute(text("SELECT 1"))
    print("Transient to pending")


listen(Session, "transient_to_pending", my_transient_to_pending)
listen(Session, "before_commit", my_before_commit)
```
The `transient_to_pending` event will throw a missing greenlet error: `sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)`
The `before_commit` event will trigger fine.

Unfortunately I haven't been able to find anything about this specific case anywhere online. Has anyone had any luck getting this to work? Thanks!

**Comment by zzzeek** [comment, marked as answer]
hi -

transient_to_pending is not called in an async context, it's called from `session.add()` which is not awaitable.  So from the architectural limitations of asyncio, there would be no way to emit IO within this method because async has to be called all the way out.

so if you are trying to do "legit" asyncio, this wouldn't be a place that you could run IO.   However, if you want to "cheat", you'd want to use the internal greenlet_spawn method directly, something like this:

```py
from sqlalchemy.util.concurrency import greenlet_spawn

async def do_stuff(async_session):
     await greenlet_spawn(async_session.add, my_object)
```

then the above `session.add` will be called with a greenlet context and the `session.execute()` inside should work

**Reply by stanleyfang** [reply]
Thanks for the quick response. I'll give the `greenlet_spawn` method a try :+1:

**Reply by CaselIT** [reply]
@zzzeek does it make sense mentioning https://github.com/sqlalchemy/awaitlet/?

**Reply by zzzeek** [reply]
awaitlet has no users at all and I dont know if I want to change that :)  but yes we can mention awaitlet here as the "public API" version of greenlet_spawn.    But I think bigger picture, someone is using asyncio that means you can't normally do IO inside of a non-asyncio method.    IMO this is still a big limitation of Python asyncio itself

**Reply by stanleyfang** [reply]
The `greenlet_spawn` approach works well for me. Thanks again!
